google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -1568,53 +1648,88 @@ public void run() {
         });
   }
 
+  /**
+   * Returns the minimum of the wanted number of sessions that the caller wants to create and the
+   * actual max number that may be created at this moment.
+   */
+  private int getAllowedCreateSessions(int wantedSessions) {
+    synchronized (lock) {
```

Comment:
isn't this sort of diverging from the initial design.. we should always enqueue waiters to get session.. the waiters can decide to get more session or not based on the max value.. this seemed to have percolated outside that creation logic.

Reply:
No, this is still the same logic as the original design, only it has been moved to a separate method as it has become a little bit more complex. It is only called before createSessions(..) is called to determine the maximum number of sessions we are allowed to create. Previously, calling `canCreateSession()` was enough to ensure that it was ok to create 1 session, now it additionally checks that it is not trying to create more than the allowed number of sessions as well.

Reply:
sgtm!